runtime/src/accounts_hash.rs

```diff
@@ -282,9 +291,18 @@ impl AccountsHash {
         // Only use the 3 level optimization if we have at least 4 levels of data.
         // Otherwise, we'll be serializing a parallel operation.
         let threshold = target * fanout;
-        let three_level = max_levels_per_pass.unwrap_or(usize::MAX) >= THREE_LEVEL_OPTIMIZATION
+        let mut three_level = max_levels_per_pass.unwrap_or(usize::MAX) >= THREE_LEVEL_OPTIMIZATION
             && total_hashes >= threshold;
-        let num_hashes_per_chunk = if three_level { target } else { fanout };
+        if three_level {
+            if let Some(specific_level_count_value) = specific_level_count {
+                three_level = specific_level_count_value >= THREE_LEVEL_OPTIMIZATION;
+            }
+        }
+        let (num_hashes_per_chunk, levels_hashed) = if three_level {
+            (target, THREE_LEVEL_OPTIMIZATION)
+        } else {
+            (fanout, 1)
+        };
```

Comment:
Can we extract this into a separate function that returns whether to `use_three_levels`

Reply:
Also could we add a few diagrams for this block of logic here: https://github.com/solana-labs/solana/blob/master/runtime/src/accounts_hash.rs#L322-L344 clarifying the structure of the logic? For me specifically, I think it's a bit hard tracking what is being iterated on over the abstracted structure underlying the generic `get_hashes()` function:

For me, a diagram of
1) The structure of what is being indexed by `i`, "data_index" "end_index", "end",
2) What are the boundaries of the values being returned by `get_hashes` on every iteration
3) The structure of the "3 levels of fanout". This seems to imply a merkle three with three layers of children and parents but I don't think that's what's being done here? 

would be super helpful!